pkg/connector/roles.go

```diff
@@ -127,44 +149,125 @@ func (o *roleBuilder) Grants(
 	annotations.Annotations,
 	error,
 ) {
-	var outputAnnotations annotations.Annotations
-	page, limit, _, err := client.ParsePaginationToken(token)
+	var bag pagination.Bag
+
+	err := bag.Unmarshal(token.Token)
 	if err != nil {
 		return nil, "", nil, err
 	}
 
-	users, total, ratelimitData, err := o.client.GetRoleUsers(
-		ctx,
-		resource.Id.Resource,
-		limit,
-		page,
-	)
-	outputAnnotations.WithRateLimiting(ratelimitData)
-	if err != nil {
-		return nil, "", outputAnnotations, err
-	}
+	state := bag.Pop()
+	if state == nil {
+		bag.Push(pagination.PageState{
+			Token:          "",
+			ResourceTypeID: userResourceType.Id,
+		})
 
-	if len(users) == 0 {
-		return nil, "", outputAnnotations, nil
+		if o.syncPermissions {
+			bag.Push(pagination.PageState{
+				Token:          "",
+				ResourceTypeID: scopeResourceType.Id,
+			})
+		}
+
+		nextToken, err := bag.Marshal()
+		if err != nil {
+			return nil, "", nil, err
+		}
+
+		return nil, nextToken, nil, nil
```

Comment:
nit: we usually do this like 
https://github.com/ConductorOne/baton-okta/blob/cf6fb6902c12a47349328033e3b5e280aa139278/pkg/connector/aws_account.go#L225-L243

which I think is cleaner